Introduced handling exceptions using Quarantine

WojciechMazur opened the pull request:
Introduced handling exceptions using Quarantine library

- `QueryBuilder.run` now returns Result[Stream[Result[T]]], since failure can accoure both on connection to Datastore and decoding of each element

- `delete` and `deleteAll` now returns Result[Unit]
- `save` and `saveAll` now returns Result[Ref[T]] / Result[Map[T, Ref[T]] accordingly  
- `Decoder[T].decodeValue` now returns `Result[T]` 

Also fixed tests to use new method results

WojciechMazur committed:
Introduced handling exceptions using Quarantine

propensive commented on src/core/MutatusException.scala:
```diff
@@ -0,0 +1,10 @@
+
+sealed abstract class MutatusException(msg: String) extends Exception(msg)
+case class SerializationException(msg: String) extends MutatusException(msg)
+case class DatabaseException(throwable: Throwable)
```
Does this have to be `Throwable`, or can it be `Exception`? I'm guessing that there are sometimes `Error`s too?

WojciechMazur replied:
It can be `Exception` as well,  I don't think it may throw `Error`s

propensive commented on src/core/mutatus.scala:
```diff
@@ -108,20 +135,23 @@ object `package` {
-    extends RuntimeException(
-      s"entity of type $kind cannot be deleted becasue it has not been saved"
-    )
+  implicit val monadicResult: Monadic[mutatus.Result] =
+    new Monadic[mutatus.Result] {
+      def flatMap[A, B](from: Result[A])(fn: A => Result[B]): Result[B] =
+        from.flatMap(fn)
+      def map[A, B](from: Result[A])(fn: A => B): Result[B] = from.map(fn)
+      def point[A](value: A): Result[A] = Result(value)
+    }
+}
```
One day, I'll move this into Quarantine itself...

propensive commented on src/core/mutatus.scala:
```diff
@@ -87,19 +103,30 @@ object `package` {
-      svc.readWrite.put(buildEntity()).getKey
-    )
+    ): mutatus.Result[Ref[T]] = Result {
+      new Ref[T](
+        svc.readWrite.put(buildEntity()).getKey
+      )
```
Will this produce one of Mutatus's exception types, or will it potentially create a `Surprise`?

WojciechMazur replied:
I think that `buildEntity` could be considered as safe operation. I don't think there is any case in which it may throw exception. I think that all possible errors would appears at compile time, e.q lack of `IdField` or lack of `Encoder[T]` 
Although `svc.readWrite.put` may result with `DatastoreException`, I'll add handling of this with `.extenuate{
      case exc: DatastoreException => DatabaseException(exc)
    }`

propensive reviewed (changes requested):
I got a bit lost further down, but I think it's best to merge this soon. Could you take a quick look at the potential issue with creating `Ref[T]`s, which I think might produce `Surprise`s? If that's all good, we can merge it.

WojciechMazur committed:
Review fixes
Handled DatastoreException at save
Changed DatabaseException to handle exception instead of throwable
